### PR TITLE
GH-35686:[Go] Add AppendTime to TimestampBuilder

### LINKED ISSUE #35686: [Go][TimestampBuilder] Add AppendTime
### Describe the enhancement requested

Right now before appending the Go `time.Time` to a `TimestampBuilder` we need to switch case the builder.type to see what is the precision. I suggest having a native `AppendTime` method.

```
func (b *TimestampBuilder) AppendTime(t time.Time) {
	switch b.dtype.Unit {
	case arrow.Nanosecond:
		b.Append(arrow.Timestamp(t.UnixNano()))
	case arrow.Microsecond:
		b.Append(arrow.Timestamp(t.UnixMicro()))
	case arrow.Millisecond:
		b.Append(arrow.Timestamp(t.UnixMilli()))
	case arrow.Second:
		b.Append(arrow.Timestamp(t.Unix()))
	default:
		panic("arrow: invalid timestamp unit")
	}
}
```

### Component(s)

Go

### PR DESCRIPTION
**Pull request by yevgenypats**
### Are there any user-facing changes?

* Closes: #35686

### CODE REVIEW

**Review by zeroshade — CHANGES_REQUESTED** [review]
Please include the updates to the tmpl files also. 

Otherwise this LGTM, though it would be good to add tests for all the unit types, not just for `Second`.

**Comment by zeroshade** [comment]
@yevgenypats Do we want to expand the tests to test more units than just `Second`? Or do you think we're fine with just testing `Second`?

**Comment by yevgenypats** [comment]
> @yevgenypats Do we want to expand the tests to test more units than just `Second`? Or do you think we're fine with just testing `Second`?

I can add but maybe we want to move TimestampBuilder out of the template as it becomes quite hard to develop inside a template and Timestamp starts to diverge from all other numeric types. wdyt?

**Comment by zeroshade** [comment]
@yevgenypats Makes sense to me, feel free to go ahead and do that.

**Comment by candiduslynx** [comment]
Hi @zeroshade!

I'm overtaking this PR from @yevgenypats.

I've extracted timestamp from numeric gen (array, builder, tests, traits).

Additionally, I propose to introduce a helper func for creation of `arrow.Timestamp` from `time.Time` and `arrow.TimeUnit`.

The logic was tested and introduced in https://github.com/cloudquery/arrow/pull/17 (in this PR: 1689eee0ec81da6e62f545eb06f8f71ed3f69a22).

**Comment by yevgenypats** [comment]
> Hi @zeroshade!
> 
> I'm overtaking this PR from @yevgenypats.
> 
> I've extracted timestamp from numeric gen (array, builder, tests, traits).
> 
> Additionally, I propose to introduce a helper func for creation of `arrow.Timestamp` from `time.Time` and `arrow.TimeUnit`.
> 
> The logic was tested and introduced in [cloudquery#17](https://github.com/cloudquery/arrow/pull/17) (in this PR: [1689eee](https://github.com/apache/arrow/commit/1689eee0ec81da6e62f545eb06f8f71ed3f69a22)).
> 
> UPD: it seems there's a [flaky](https://github.com/apache/arrow/actions/runs/5077433733/jobs/9120706233?pr=35687#step:5:739) test (`go/arrow/flight/flightsql/driver/driver_test.go:741`, `TestPreparedStatementNoSchema`)

Thanks @candiduslynx !

File: go/arrow/type_traits_test.go [hunk 1]
```diff
@@ -233,3 +233,37 @@ func TestMonthDayNanoIntervalTraits(t *testing.T) {
...
+	})
+
+	b2 := make([]byte, arrow.TimestampTraits.BytesRequired(N))
+	for i := 0; i < N; i++ {
+		beg := i * arrow.TimestampSizeBytes
+		end := (i + 1) * arrow.TimestampSizeBytes
+		arrow.TimestampTraits.PutValue(b2[beg:end], arrow.Timestamp(i))
+	}
+
+	if !reflect.DeepEqual(b1, b2) {
```
[the remark this answers is not in the document]
Comment (candiduslynx) [reply go/arrow/type_traits_test.go:250]: I just copy-pasted from the generated code. I can fix it here, if you insist.

[the remark this answers is not in the document]
Comment (candiduslynx) [reply go/arrow/type_traits_test.go]: Done in e57e2ae596a072ec3e0d1da6460f7d21e86cbd55

**Review by zeroshade — APPROVED** [review]
LGTM Thanks!

**Pull request merged**